Integrations/integration-ArcSightESM.yml

```diff
@@ -56,1158 +56,946 @@ configuration:
-        }
-        if (!response) {
-            throw 'Events response to query is empty';
-        }
-        var fieldNames = response.columnHeaders;
-        if (fieldNames.indexOf(timeField) === -1) {
-            throw 'Events Query must contain ' + timeField + ' fields';
-        }
-        var data = response.data;
-        var lastEventTime = lastRun && lastRun.lastEventTime ? lastRun.lastEventTime : 0;
-        var maxEventTime = lastEventTime;
-        for (var i = 0; i < data.length; i++) {
-            var eventTimeInMillis = Number(data[i][timeField]);
-            if (lastEventTime < eventTimeInMillis) {
-                if (maxEventTime < eventTimeInMillis) {
-                    maxEventTime = eventTimeInMillis;
-                }
-
-                var keys = Object.keys(data[i]);
-                var labels = [];
-                for (var j = 0; j < keys.length; j++) {
-                    if (keys[j] !== 'Name') {
-                        labels.push({
-                            type: keys[j],
-                            value: String(data[i][keys[j]])
-                        });
-                    }
+        """
+        results = [{field: result.get('value')[idx].get('$') for idx, field in enumerate(fields)} for result in results]
+        return fields, results
+
+
+    def get_query_viewer_results_command():
+        resource_id = demisto.args().get('ids')
+        columns, query_results = get_query_viewer_results(query_viewer_id=resource_id)
+
+        demisto.results({
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['json'],
+            'Contents': {'results': query_results}
+        })
+
+
+    def fetch():
+        """
+        Query viewer should be defined in ArcSight ESM. fetch incidents fetches the results of query viewer
+        and converts them to Demisto incidents. We can query Cases or Events. If Cases are fetched then the
+        query viewer query must return fields ID and Create Time. If Events are fetched then Event ID and Start Time.
+        """
+        events_query_viewer_id = demisto.params().get('events_query_viewer_id')
+        cases_query_viewer_id = demisto.params().get('cases_query_viewer_id')
+
+        last_run = demisto.getLastRun()
+        last_create_time = last_run.get('last_create_time', 0)
+        already_fetched = deque(last_run.get('already_fetched', []), maxlen=1000)
+        latest_created_time = last_create_time
+
+        fields, query_results = get_query_viewer_results(events_query_viewer_id or cases_query_viewer_id)
+        incidents = []
+        for result in query_results:
+            # convert case or event to demisto incident
+            r_id = result.get('ID') or result.get('Event ID')
+            create_time = int(result.get('Start Time') or result.get('Create Time'))
+            if create_time >= last_create_time and r_id not in already_fetched:
```

Comment:
you should have two variables for time.
last_create_time => you will compare with create_time (no one said that arcsight will return the cases in ascending or descending order)
next_last_create_time => this will you update in each iteration with the latest case/event time

the condition should be `if create_time >= last_create_time or r_id not in already_fetched:` `or` and not `and`

Reply:
`if create_time >= last_create_time or r_id not in already_fetched`
wouldn't this make fetching duplicates possible? 
regarding time variables, I think we should go over this on Sunday.

Reply:
also, in your code:
> if create_time < last_create_time:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;continue
>
>\# check if case id already was fetched before
>\# if yes, then don't create incident from that case
>case_id = case['ID']
>incident_exists = False
>for existing_case_id in last_case_ids:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;    if existing_case_id == case_id:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        incident_exists = True
>
>if incident_exists:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;continue

which translates into:
<pre><code>if create_time < last_create_time <b>or</b> case_id in last_case_ids:</pre></code>
so the contrapositive will be:
<pre><code>if create_time => last_create_time <b>and</b> case_id not in last_case_ids:</pre></code>

Reply:
lets go over it on Sunday